### PR TITLE
docs: Add deployment documentation

### PR DESCRIPTION
**Pull request by orelmisan**
### What this PR does
Before this PR:

After this PR:
Add a document to list the objects that are deployed when a user / operator deploys KubeVirt.

Since these objects can change, the documentation contains a `kubectl get` command in order to view each object in detail.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
This document excludes most of the objects related to monitoring and objects that are OKD/OpenShift specific.
A follow-up PR could also add diagrams depicting the relations between the objects.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: docs/deployment.md [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+A dedicated [Namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/) is created for KubeVirt's components.
+
+```shell
+kubectl get namespace kubevirt -o yaml
+```
+
+## KubeVirt CustomResourceDefinition
+KubeVirt extends Kubernetes using [CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) (CRDs).
+
+The first deployed CRD is `kubevirts.kubevirt.io` which allows us to later create the `KubeVirt` CustomResource.
```
Comment (codingben) [inline docs/deployment.md:37]: Hi Orel, this is great idea! I think it would be longer than usual to read, maybe you can make table instead? Similar to what we have in SSP docs [1].

[1] https://github.com/kubevirt/ssp-operator/blob/main/docs/installation.md#kubernetes

Comment (orelmisan) [reply docs/deployment.md:37]: I thought about it in the beginning, and came to the conclusion that the `kubectl get` commands will break.

Reading the document from start to finish is indeed not trivial, but using an IDE (or a similar tool that helps read markdown files) is more convenient - since it provides a generated table of content.

Comment (orelmisan) [reply docs/deployment.md:37]: Additionally, there is a feature in GitHub to view the TOC automatically:
https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/

Unfortunately it only works for merged files.

File: docs/deployment.md [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+
+## Secrets
+```shell
+kubectl get --namespace=kubevirt secret kubevirt-ca -o yaml
+kubectl get --namespace=kubevirt secret kubevirt-export-ca -o yaml
+```
+
+## ClusterInstancetypes
+```shell
+ ubectl get virtualmachineclusterinstancetypes.instancetype.kubevirt.io -o yaml
```
Comment (codingben) [inline docs/deployment.md:472]: `kubectl`

File: docs/deployment.md (hunk 1 above)
Comment (codingben) [reply docs/deployment.md:37]: Okay. What about this suggestion to make it shorter:

## Export Proxy Resources

Not deployed by default. Deployed only when `VMExport` feature gate is added to KubeVirt CR.

```shell
# Cluster Resources
kubectl get clusterroles.rbac.authorization.k8s.io kubevirt-exportproxy -o yaml
kubectl get clusterrolebindings.rbac.authorization.k8s.io kubevirt-exportproxy -o yaml

# Namespace Resources
kubectl get --namespace=kubevirt serviceaccount kubevirt-exportproxy -o yaml
kubectl get --namespace=kubevirt roles.rbac.authorization.k8s.io  kubevirt-exportproxy -o yaml
kubectl get --namespace=kubevirt rolebindings.rbac.authorization.k8s.io kubevirt-exportproxy -o yaml
kubectl get --namespace=kubevirt secret kubevirt-exportproxy-certs -o yaml
kubectl get --namespace=kubevirt deployments.apps virt-exportproxy -o yaml
kubectl get --namespace=kubevirt service virt-exportproxy -o yaml
kubectl get --namespace=kubevirt configmap kubevirt-ca -o yaml
kubectl get --namespace=kubevirt configmap kubevirt-export-ca -o yaml
kubectl get --namespace=kubevirt secret kubevirt-ca -o yaml
kubectl get --namespace=kubevirt secret kubevirt-export-ca -o yaml
```

File: docs/deployment.md (hunk 2 above)
Comment (orelmisan) [reply docs/deployment.md]: Good catch!
Done.

File: docs/deployment.md (hunk 1 above)
Comment (orelmisan) [reply docs/deployment.md:37]: Interesting suggestion, I'll consider it.
The issue with it is that its harder to insert explanations between the list of resources.

File: docs/deployment.md [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+kubectl get namespace kubevirt -o yaml
+```
+
+## KubeVirt CustomResourceDefinition
+KubeVirt extends Kubernetes using [CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) (CRDs).
+
+The first deployed CRD is `kubevirts.kubevirt.io` which allows us to later create the `KubeVirt` CustomResource.
+
+```shell
+kubectl get customresourcedefinitions.apiextensions.k8s.io kubevirts.kubevirt.io -o yaml
```
Comment (jean-edouard) [inline docs/deployment.md:40]: For readability, I think we should drop all the *.k8s.io suffixes, like:

```suggestion
kubectl get customresourcedefinitions kubevirts.kubevirt.io -o yaml
```

File: docs/deployment.md [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+Binds the `kubevirt-operator` ClusterRole to the `kubevirt:kubevirt-operator` ServiceAccount.
+
+```shell
+kubectl get clusterrolebindings.rbac.authorization.k8s.io kubevirt-operator -o yaml
+```
+
+## Role
+
+```shell
+kubectl get --namespace=kubevirt roles.rbac.authorization.k8s.io kubevirt-operator -o yaml
```
Comment (jean-edouard) [inline docs/deployment.md:75]: I'd also use singular instead of plural when getting a specific object and, like `-o`, I feel that `-n` is widespread enough to avoid using the long version.
```suggestion
kubectl get role kubevirt-operator -n kubevirt -o yaml
```
That reads almost like normal text, i.e. "get the role named kubevirt-operator in the kubevirt namespace".

I realize this is very nit-picky and I wont insist on any of it, it's really just personal preference.

**Review by jean-edouard — COMMENTED** [review]
/approve
Just some suggestions in-line. Thank you!

File: docs/deployment.md [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+
+## KubeVirt CustomResource
+
+After the `virt-operator` and its dependencies are deployed, we deploy the KubeVirt CustomResource in order to trigger virt-operator to reconcile it and deploy KubeVirt's main components.
+
+```shell
+kubectl get --namespace=kubevirt kubevirts.kubevirt.io kubevirt -o yaml
+```
+
+# Operands
```
Comment (fabiand) [inline docs/deployment.md:100]: Wow.

How will you ensure that all the stuff below will be kept up to date?

File: docs/deployment.md [hunk 6]
```diff
@@ -0,0 +1,483 @@
+KubeVirt Deployment
```
Comment (fabiand) [inline docs/deployment.md:1]: What is the purpose of this document? Maybe worth adding it to the document itself.

File: docs/deployment.md [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+## KubeVirt CustomResourceDefinition
+KubeVirt extends Kubernetes using [CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) (CRDs).
+
+The first deployed CRD is `kubevirts.kubevirt.io` which allows us to later create the `KubeVirt` CustomResource.
+
+```shell
+kubectl get customresourcedefinitions.apiextensions.k8s.io kubevirts.kubevirt.io -o yaml
+```
+
+## Priority Class
```
Comment (fabiand) [inline docs/deployment.md:43]: Why is all of the below layed out here? What is the value/benefit?

Couldn't we rather get this from the operator?

**Review by fabiand — COMMENTED** [review]
I'm owrried that the doc will quickly be outdated.

I also wonder if it's not better to get this from the operator (to get from the operator what the operator will be doing). i.e.

```console
$ virt-operator --dry-run …
POST nnn
POST mmm
…
…
```

File: docs/deployment.md (hunk 4 above)
Comment (orelmisan) [reply docs/deployment.md:75]: Done.

File: docs/deployment.md (hunk 3 above)
Comment (orelmisan) [reply docs/deployment.md:40]: Removed most of the suffixes, except for the more advanced resources:
- priorityclasses.scheduling.k8s.io
- validatingwebhookconfigurations.admissionregistration.k8s.io
- mutatingwebhookconfigurations.admissionregistration.k8s.io
- apiservices.apiregistration.k8s.io

File: docs/deployment.md (hunk 6 above)
Comment (orelmisan) [reply docs/deployment.md:1]: You are absolutely right.
Done.

File: docs/deployment.md (hunk 5 above)
Comment (orelmisan) [reply docs/deployment.md:99]: The reasons for updating this document could be:
1. Renaming one or more objects.
2. Addition or removal of objects (such as a CRD).

The document intentionally does not contain any details about the objects and instructs the user to fetch the details by name using the `kubectl` utility.

There is a checkbox in the PR's TODO list, it currently refers to the user-guide but could also refer to the developer's guide.
WDYT?

File: docs/deployment.md (hunk 7 above)
Comment (orelmisan) [reply docs/deployment.md:45]: Security professionals could also gain better understanding of the RBAC rules used by the different components and audit them.

**Comment by orelmisan** [comment]
@fabiand Could you please have another look?

**Comment by fabiand** [comment]
I did.

So, why do we repeat all the steps that the operator is doing?
Who wil be keeping this up to date?
Who is benefiting from this change?

I still wonder about the value and maintainability.

**Comment by fabiand** [comment]
> 1.    Its easier to understand the full picture from the text rather from the code. A follow-up PR could add diagrams to better describe the project's structure.

Right. But this could be automated by letting virt-operator simply spit this out, maybe even with a diagram!

> 2.    It exposes the results of virt-operator deployment process to a wider audience - for example people who are proficient in Kubernetes but are not developers or are not proficient in Go.

Having virt-operator to dump all steps would also address this

> 3.    It exposes the structure of the KubeVirt project in a more detailed way, which may benefit new contributors and existing contributors who want to expand their knowledge and go wider or deeper.

To me one flaw in this - and the virt operator dumping it's steps - is that you have to have muhc knowgedle to understand all steps and understand their impact.

> 4.    It exposes (some of) the Kubernetes extension mechanisms that are used by KubeVirt: CRDs, API aggregation, webhooks.

A virt-operator --dump-steps approach  would do this as well
FWIW I don't udnerstand the value of this

> 5.    It exposes all the RBAC rules used by KubeVirt for security professionals to audit.

This is a relevant topic. But here a static doc is bad: It can be outdated.
I would lve to see a tool which is dumping the aggregated RBAC permissions of kubevirt.
To me this doc is not the answer to this problem.

tl;dr

- this static page is just adding more stuff to maintain
- there is an alternative by making virt-operator dump all of this
- Understanding RBAC should deserve a dedicated tool.
